File: backend/src/main/java/com/example/backend/auth/AuthService.java
package com.example.backend.auth;

import com.example.backend.Security.JwtService;
import com.example.backend.dto.*;
import com.example.backend.entity.*;
import com.example.backend.repository.*;
import com.example.backend.auth.EmailService; // Импортируем наш сервис писем
import lombok.RequiredArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.Optional;
import java.util.Random;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final TenantProfileRepository tenantProfileRepository;
    private final LandlordProfileRepository landlordProfileRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuthenticationManager authenticationManager;
    private final EmailService emailService; // Подключаем отправку email

    @Transactional
    public AuthResponse register(RegisterRequest request) {

        // 1. Проверяем email на статус ACTIVE или очищаем старый мусор
        Optional<User> existingUserByEmail = userRepository.findByEmail(request.getEmail());
        if (existingUserByEmail.isPresent()) {
            User user = existingUserByEmail.get();
            if (user.getStatus() == UserStatus.ACTIVE) {
                throw new RuntimeException("Пользователь с таким email уже существует");
            } else {
                // Пользователь не подтвердил почту. Удаляем старую запись.
                userRepository.delete(user);
                userRepository.flush(); // Принудительно применяем удаление в БД прямо сейчас
            }
        }

        // 2. Проверяем ИНН (чтобы не было ошибки 500 при смене почты, но старом ИНН)
        if (request.getRole() == Role.TENANT) {
            Optional<TenantProfile> existingProfile = tenantProfileRepository.findByInn(request.getInn());
            if (existingProfile.isPresent()) {
                if (existingProfile.get().getUser().getStatus() == UserStatus.ACTIVE) {
                    throw new RuntimeException("Арендатор с таким ИНН уже зарегистрирован");
                } else {
                    userRepository.delete(existingProfile.get().getUser());
                    userRepository.flush();
                }
            }
        } else if (request.getRole() == Role.LANDLORD) {
            Optional<LandlordProfile> existingProfile = landlordProfileRepository.findByInn(request.getInn());
            if (existingProfile.isPresent()) {
                if (existingProfile.get().getUser().getStatus() == UserStatus.ACTIVE) {
                    throw new RuntimeException("Арендодатель с таким ИНН уже зарегистрирован");
                } else {
                    userRepository.delete(existingProfile.get().getUser());
                    userRepository.flush();
                }
            }
        }

        // 3. Создаем нового пользователя
        String code = String.format("%06d", new Random().nextInt(1000000));

        User user = User.builder()
                .email(request.getEmail())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .role(request.getRole())
                .status(UserStatus.UNVERIFIED)
                .verificationCode(code)
                .codeExpiresAt(LocalDateTime.now().plusMinutes(2))
                .build();

        User savedUser = userRepository.save(user);

        // 4. Создаем профили
        if (request.getRole() == Role.TENANT) {
            TenantProfile tenantProfile = TenantProfile.builder()
                    .user(savedUser)
                    .name(request.getName())
                    .inn(request.getInn())
                    .phone(request.getPhone())
                    .build();
            tenantProfileRepository.save(tenantProfile);

        } else if (request.getRole() == Role.LANDLORD) {
            LandlordProfile landlordProfile = LandlordProfile.builder()
                    .user(savedUser)
                    .companyName(request.getName())
                    .inn(request.getInn())
                    .phone(request.getPhone())
                    .isVerified(false)
                    .build();
            landlordProfileRepository.save(landlordProfile);
        }

        // 5. Отправляем письмо
        emailService.sendVerificationCode(savedUser.getEmail(), code);

        return AuthResponse.builder()
                .message("Код подтверждения отправлен на вашу почту.")
                .build();
    }

    @Transactional
    public AuthResponse verifyEmail(VerifyEmailRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));

        if (user.getStatus() == UserStatus.ACTIVE) {
            throw new RuntimeException("Email уже подтвержден");
        }
        if (user.getCodeExpiresAt().isBefore(LocalDateTime.now())) {
            throw new RuntimeException("Код истек. Запросите новый.");
        }
        if (!user.getVerificationCode().equals(request.getCode())) {
            throw new RuntimeException("Неверный код");
        }

        // Успешная верификация
        user.setStatus(UserStatus.ACTIVE);
        user.setVerificationCode(null);
        user.setCodeExpiresAt(null);
        userRepository.save(user);

        // Теперь выдаем токен
        String jwtToken = jwtService.generateToken(user);
        return AuthResponse.builder()
                .token(jwtToken)
                .email(user.getEmail())
                .role(user.getRole())
                .message("Успешная авторизация")
                .build();
    }

    @Transactional
    public AuthResponse resendCode(String email) {
        User user = userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));

        if (user.getStatus() == UserStatus.ACTIVE) {
            throw new RuntimeException("Email уже подтвержден");
        }

        // Проверка на Кулдаун: если время истечения еще в будущем, значит 2 минуты не прошло
        if (user.getCodeExpiresAt() != null && user.getCodeExpiresAt().isAfter(LocalDateTime.now())) {
            throw new RuntimeException("Вы можете запросить новый код только через 2 минуты");
        }

        String newCode = String.format("%06d", new Random().nextInt(1000000));
        user.setVerificationCode(newCode);
        user.setCodeExpiresAt(LocalDateTime.now().plusMinutes(2));
        userRepository.save(user);

        emailService.sendVerificationCode(user.getEmail(), newCode);

        return AuthResponse.builder().message("Новый код отправлен").build();
    }

    public AuthResponse login(LoginRequest request) {
        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(request.getEmail(), request.getPassword())
        );

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("Пользователь не найден"));

        // Не даем войти, если почта не подтверждена
        if (user.getStatus() == UserStatus.UNVERIFIED) {
            throw new RuntimeException("Пожалуйста, подтвердите вашу электронную почту");
        }

        String jwtToken = jwtService.generateToken(user);
        return AuthResponse.builder()
                .token(jwtToken)
                .email(user.getEmail())
                .role(user.getRole())
                .message("Успешная авторизация")
                .build();
    }
}

File: frontend/lib/service/auth_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

class AuthService {
// 1. Оставляем в baseUrl ТОЛЬКО хост и порт (без /api)
final Dio _dio = Dio(BaseOptions(baseUrl: 'http://127.0.0.1:8080'));
final FlutterSecureStorage _storage = const FlutterSecureStorage();

Future<bool> login(String email, String password) async {
  // 1. Очищаем от случайных пробелов в начале и конце
  final cleanEmail = email.trim();
  final cleanPassword = password.trim();

  // 2. Давай выведем их в консоль в кавычках, чтобы визуально убедиться,
  // что внутри нет переносов строк или других скрытых символов
  print('Пытаемся войти...');
  print('Email: "$cleanEmail"');
  print('Password: "$cleanPassword"');

  try {
    final response = await _dio.post('/api/auth/login', data: {
      'email': cleanEmail, // Отправляем очищенные данные
      'password': cleanPassword,
    });

      if (response.statusCode == 200) {
        final token = response.data['token'];
        final role = response.data['role'];

        await _storage.write(key: 'jwt_token', value: token);
        await _storage.write(key: 'user_role', value: role);

        return true;
      }
      return false;
    } on DioException catch (e) {
      // Выводим статус-код и принудительно показываем сообщение Dio, если данных нет
      final statusCode = e.response?.statusCode;
      final serverData = e.response?.data;

      print('Ошибка входа! Код: $statusCode');
      print('Ответ сервера: "${serverData}"');
      print('Системная ошибка Dio: ${e.message}');

      return false;
    } catch (e) {
      print('Неизвестная ошибка: $e');
      return false;
    }
  }
// Регистрация
  Future<bool> register(String email, String password, String role, String name, String inn, String phone) async {
    final cleanEmail = email.trim();
    final cleanPassword = password.trim();
    try {
      final response = await _dio.post('/api/auth/register', data: {
        'email': cleanEmail,
        'password': cleanPassword,
        'role': role.trim(), // 'TENANT' или 'LANDLORD'
        'name': name.trim(),
        'inn': inn.trim(),
        'phone': phone.trim(),
      });
      return response.statusCode == 200;
    } catch (e) {
      print('Ошибка регистрации: $e');
      return false;
    }
  }

  // Подтверждение кода
  Future<bool> verifyEmail(String email, String code) async {
    final cleanEmail = email.trim();
    try {
      final response = await _dio.post('/api/auth/verify', data: {
        'email': cleanEmail,
        'code': code,
      });

      if (response.statusCode == 200) {
        // Сохраняем токен, так как бэкенд выдает его после верификации
        final token = response.data['token'];
        final role = response.data['role'];
        await _storage.write(key: 'jwt_token', value: token);
        await _storage.write(key: 'user_role', value: role);
        return true;
      }
      return false;
    } catch (e) {
      print('Ошибка верификации: $e');
      return false;
    }
  }

  // Переотправка кода
  Future<bool> resendCode(String email) async {
    final cleanEmail = email.trim();
    try {
      final response = await _dio.post('/api/auth/resend-code', queryParameters: {
        'email': cleanEmail,
      });
      return response.statusCode == 200;
    } catch (e) {
      print('Ошибка переотправки: $e');
      return false;
    }
  }
  Future<String?> getToken() async {
    return await _storage.read(key: 'jwt_token');
  }
}